compiler/rustc_codegen_ssa/src/sir.rs

```diff
@@ -890,3 +890,136 @@ impl SirTypes {
         })
     }
 }
+
+pub mod labels {
+    use object::{Object, ObjectSection};
+    use std::{
+        convert::TryFrom,
+        fs,
+        path::{Path, PathBuf},
+        process::Command,
+    };
+
+    fn split_symbol(s: &str) -> (String, u32) {
+        let data: Vec<&str> = s.split(':').collect();
+        debug_assert!(data.len() == 3);
+        let sym = data[1].to_owned();
+        let bb_idx = data[2].parse::<u32>().unwrap();
+        (sym, bb_idx)
+    }
+
+    pub fn add_yk_label_section(exe_path: &Path) {
+        let labels = extract_dwarf_labels(exe_path).unwrap();
+        let mut tempf = tempfile::NamedTempFile::new().unwrap();
+        bincode::serialize_into(&mut tempf, &labels).unwrap();
+        add_section(exe_path, tempf.path());
+    }
+
+    fn add_section(exe_path: &Path, sec_data_path: &Path) {
+        let mut out_path = PathBuf::from(exe_path);
+        out_path.set_extension("with_labels");
+        Command::new("objcopy")
+            .args(&[
+                "--add-section",
+                &format!("{}={}", ykpack::YKLABELS_SECTION, sec_data_path.to_str().unwrap()),
+                "--set-section-flags",
+                &format!("{}=contents,alloc,readonly", ykpack::YKLABELS_SECTION),
+                exe_path.to_str().unwrap(),
+                out_path.to_str().unwrap(),
+            ])
+            .output()
+            .expect("failed to insert labels section");
+        std::fs::rename(out_path, exe_path).unwrap();
+    }
+
+    fn extract_dwarf_labels(exe_filename: &Path) -> Result<Vec<ykpack::SirLabel>, gimli::Error> {
+        let file = fs::File::open(exe_filename).unwrap();
+        let mmap = unsafe { memmap::Mmap::map(&file).unwrap() };
+        let object = object::File::parse(&*mmap).unwrap();
+        let endian = if object.is_little_endian() {
+            gimli::RunTimeEndian::Little
+        } else {
+            gimli::RunTimeEndian::Big
+        };
+        let mut labels = Vec::new();
+
+        let loader = |id: gimli::SectionId| -> Result<&[u8], gimli::Error> {
+            Ok(object
+                .section_by_name(id.name())
+                .map(|sec| sec.data().expect("failed to decompress section"))
+                .unwrap_or(&[] as &[u8]))
+        };
+        let sup_loader = |_| Ok(&[] as &[u8]);
+        let dwarf_cow = gimli::Dwarf::load(&loader, &sup_loader)?;
+        let borrow_section: &dyn for<'a> Fn(
+            &&'a [u8],
+        )
+            -> gimli::EndianSlice<'a, gimli::RunTimeEndian> =
+            &|section| gimli::EndianSlice::new(section, endian);
+        let dwarf = dwarf_cow.borrow(&borrow_section);
+        let mut iter = dwarf.units();
+        let mut subaddr = None;
+        while let Some(header) = iter.next()? {
```

Comment:
Is this just:

```rust
for header in &iter {
   ...
}
```

in disguise?

Reply:
I have a feeling there's a reason for that. Will check.

Reply:
Yes.

The reason for `while let Some` is that [these iterators are not standard](https://docs.rs/gimli/0.23.0/gimli/read/struct.DebugInfoUnitHeadersIter.html). They return `Result`:
![image](https://user-images.githubusercontent.com/604955/101369495-4c3e3480-38a0-11eb-97f4-0084dad4b8c2.png)

So that means you can't use a for loop.

I doubled checked the [crate examples](https://github.com/gimli-rs/gimli/blob/d36fcf11c878c8ebfb4da4cad300773869b5b5e4/examples/simple.rs#L48), and they use `while let` too.

Looking at the features of `object` there's a [`std` feature](https://github.com/gimli-rs/gimli/blob/master/Cargo.toml#L45) that pulls in `fallible-iterator` which might let us use a for loop, but is it worth it?

Reply:
I leave that up to you :)